From 4d0d7646747d009333974a68cab104aadd4f9d4e Mon Sep 17 00:00:00 2001
Message-Id: <4d0d7646747d009333974a68cab104aadd4f9d4e.1650636482.git.keller@fs-net.de>
In-Reply-To: <8876d3464f7abe86d96420dca69dd3317368c208.1650636480.git.keller@fs-net.de>
References: <8876d3464f7abe86d96420dca69dd3317368c208.1650636480.git.keller@fs-net.de>
From: Lihua Liu <lihual@codeaurora.org>
Date: Tue, 17 Nov 2020 14:49:09 +0800
Subject: [PATCH 45/56] qcacld-2.0: Fix a potential overflow issue

Overflow may happen in function ol_tx_vdev_set_bundle_require
while calculating vdev->bundling_reqired as both high_th and
low_th are configurable in range of {0, 70000}, which will
cause false triggering for bundle. To address this, change
variable high_th/low_th to uint64_t.

Change-Id: Ic52373c3e91a5e3c4fa7067f0336a5f5dd0a35bb
CRs-Fixed: 2816823
---
 CORE/CLD_TXRX/TXRX/ol_tx.c              | 2 +-
 CORE/SERVICES/COMMON/ol_txrx_osif_api.h | 6 +++---
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/CORE/CLD_TXRX/TXRX/ol_tx.c b/CORE/CLD_TXRX/TXRX/ol_tx.c
index 0555eee..662fa71 100644
--- a/CORE/CLD_TXRX/TXRX/ol_tx.c
+++ b/CORE/CLD_TXRX/TXRX/ol_tx.c
@@ -1987,7 +1987,7 @@ ol_tx_pdev_reset_bundle_require(void* pdev_handle)
  */
 void
 ol_tx_vdev_set_bundle_require(uint8_t vdev_id, unsigned long tx_bytes,
-			uint32_t time_in_ms, uint32_t high_th, uint32_t low_th)
+			uint32_t time_in_ms, uint64_t high_th, uint64_t low_th)
 {
 	struct ol_txrx_vdev_t* vdev = ol_txrx_get_vdev_from_vdev_id(vdev_id);
 	bool old_bundle_required;
diff --git a/CORE/SERVICES/COMMON/ol_txrx_osif_api.h b/CORE/SERVICES/COMMON/ol_txrx_osif_api.h
index b770729..029ec34 100644
--- a/CORE/SERVICES/COMMON/ol_txrx_osif_api.h
+++ b/CORE/SERVICES/COMMON/ol_txrx_osif_api.h
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2012, 2014, 2017-2018 The Linux Foundation. All rights reserved.
+ * Copyright (c) 2012, 2014, 2017-2018, 2020 The Linux Foundation. All rights reserved.
  *
  * Previously licensed under the ISC license by Qualcomm Atheros, Inc.
  *
@@ -255,14 +255,14 @@ ol_tx_pdev_reset_driver_del_ack(void *pdev)
 #ifdef QCA_SUPPORT_TXRX_HL_BUNDLE
 void
 ol_tx_vdev_set_bundle_require(uint8_t vdev_id, unsigned long tx_packets,
-			uint32_t time_in_ms, uint32_t high_th, uint32_t low_th);
+			uint32_t time_in_ms, uint64_t high_th, uint64_t low_th);
 void
 ol_tx_pdev_reset_bundle_require(void* pdev);
 
 #else
 static inline void
 ol_tx_vdev_set_bundle_require(uint8_t vdev_id, unsigned long tx_packets,
-			uint32_t time_in_ms, uint32_t high_th, uint32_t low_th)
+			uint32_t time_in_ms, uint64_t high_th, uint64_t low_th)
 {
 	return;
 }
-- 
2.14.5

